hail/src/main/scala/is/hail/types/Box.scala

```diff
@@ -1,19 +1,25 @@
 package is.hail.types
 
-final case class Box[T](var b: Option[T] = None, matchCond: (T, T) => Boolean = { (a: T, b: T) => a == b }) {
-  def unify(t: T): Boolean = b match {
+import java.util.function._
+
+final case class Box[T](
+  b: ThreadLocal[Option[T]] = ThreadLocal.withInitial(
+    new Supplier[Option[T]] { def get = None }),
```

Comment:
Ugh, this is real gross.  I guess the alternative is for `unify` to take and populate a `mutable.Map[TVariable, Type]`.  I guess I prefer that, but this should be fine as a first step.

Reply:
I'll have you take this up with the compiler's team. 📞  @tpoterba 

(Tim: I managed to avoid making the entire function registry thread local but just making the Box's `b` thread local, but Cotton is still disgusted. 😉 )